modin/experimental/batch/pipeline.py

```diff
@@ -0,0 +1,396 @@
+    is_output : bool, default: False
+        Whether this query is an output query and should be passed both to the next query, and
+        directly to postprocessing.
+    repartition_after : bool, default: False
+        Whether to repartition after this query is computed. Currently, repartitioning is only
+        supported if there is 1 partition prior to repartitioning.
+    fan_out : bool, default: False
+        Whether to fan out this node. If True and only 1 partition is passed as input, the partition
+        is replicated `PandasQueryPipeline.num_partitions` (default: `NPartitions.get`) times, and
+        the function is called on each. The `reduce_fn` must also be specified.
+    pass_partition_id : bool, default: False
+        Whether to pass the numerical partition id to the query.
+    reduce_fn : Callable, default: None
+        The reduce function to apply if `fan_out` is set to True. This takes the
+        `PandasQueryPipeline.num_partitions` (default: `NPartitions.get`) partitions that result from
+        this query, and combines them into 1 partition.
+    output_id : int, default: None
+            An id to assign to this node if it is an output.
+
+    Notes
+    -----
+    `func` must be a function that is applied along an axis of the dataframe.
+
+    Use `pandas` for any module level functions inside `func` since it operates directly on
+    partitions.
+    """
+
+    def __init__(
+        self,
+        func: Callable,
+        is_output: bool = False,
+        repartition_after: bool = False,
+        fan_out: bool = False,
+        pass_partition_id: bool = False,
+        reduce_fn: Optional[Callable] = None,
+        output_id: Optional[int] = None,
+    ):
+        self.func = func
+        self.is_output = is_output
+        self.repartition_after = repartition_after
+        self.fan_out = fan_out
+        self.pass_partition_id = pass_partition_id
+        self.reduce_fn = reduce_fn
+        self.output_id = output_id
+        # List of sub-queries to feed into this query, if this query is an output node.
+        self.operators = None
+
+
+class PandasQueryPipeline(object):
+    """
+    Internal representation of a query pipeline.
+
+    This object keeps track of the functions that compose to form a query pipeline.
+
+    Parameters
+    ----------
+    df : modin.pandas.Dataframe
+        The dataframe to perform this pipeline on.
+    num_partitions : int, optional
+        The number of partitions to maintain for the batched dataframe.
+        If not specified, the value is assumed equal to ``NPartitions.get()``.
+
+    Notes
+    -----
+    Only row-parallel pipelines are supported. All queries will be applied along the row axis.
+    """
+
+    def __init__(self, df, num_partitions: Optional[int] = None):
+        if Engine.get() != "Ray" or (
+            not isinstance(df._query_compiler._modin_frame, PandasOnRayDataframe)
+        ):  # pragma: no cover
```

Comment:
```suggestion
        if get_current_execution() != "PandasOnRay":
```

Reply:
Updated! I'd like to keep the check that the `_modin_frame` is backed by `PandasOnRayDataframe` as @devin-petersohn suggested.